pkg/sbproxy/sbproxy.go

```diff
@@ -0,0 +1,234 @@
+	"time"
+
+	"net/http"
+
+	"github.com/Peripli/service-broker-proxy/pkg/logger"
+	"github.com/Peripli/service-broker-proxy/pkg/osb"
+	"github.com/Peripli/service-broker-proxy/pkg/platform"
+	"github.com/Peripli/service-broker-proxy/pkg/sbproxy/middleware"
+	"github.com/Peripli/service-broker-proxy/pkg/sbproxy/server"
+	"github.com/Peripli/service-broker-proxy/pkg/sm"
+	"github.com/gorilla/mux"
+	"github.com/onrik/logrus/filename"
+	"github.com/onrik/logrus/formatter"
+	"github.com/pkg/errors"
+	"github.com/pmorie/osb-broker-lib/pkg/metrics"
+	"github.com/pmorie/osb-broker-lib/pkg/rest"
+	osbserver "github.com/pmorie/osb-broker-lib/pkg/server"
+	prom "github.com/prometheus/client_golang/prometheus"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// BrokerPathParam for the broker id
+	BrokerPathParam = "brokerID"
+
+	// APIPrefix for the Proxy OSB API
+	APIPrefix = "/v1/osb"
+
+	// Path for the Proxy OSB API
+	Path = APIPrefix + "/{" + BrokerPathParam + "}"
+)
+
+var (
+	group sync.WaitGroup
+)
+
+// SBProxy type is the starting point of the proxy application. It glues the proxy REST API and the timed
+// jobs for broker registrations
+type SBProxy struct {
+	CronScheduler *cron.Cron
+	Server        *osbserver.Server
+	AppConfig     *server.AppConfiguration
+}
+
+// New builds a new SBProxy from the provided configuration using the provided platform client. The
+// platform client is used by the SBProxy to call to the platform during broker creation and deletion.
+func New(config *Configuration, client platform.Client) (*SBProxy, error) {
+
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	appConfig := config.App
+	setUpLogging(appConfig.LogLevel, appConfig.LogFormat)
+
+	osbServer, err := defaultOSBServer(config.Osb)
+	if err != nil {
+		return nil, err
+	}
+
+	osbServer.Router.Use(middleware.LogRequest())
+
+	cronScheduler := cron.New()
+
+	regJob, err := defaultRegJob(&group, client, config.Sm, config.App.Host+APIPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cronScheduler.AddJob("@every 1m", regJob); err != nil {
+		return nil, errors.Wrap(err, "error adding registration job")
+	}
+
+	return &SBProxy{
+		Server:        osbServer,
+		CronScheduler: cronScheduler,
+		AppConfig:     appConfig,
+	}, nil
+}
+
+// Use provides a way to plugin middleware in the SBProxy
+func (s SBProxy) Use(middleware func(handler http.Handler) http.Handler) {
+	s.Server.Router.Use(middleware)
+}
+
+// Run is the entrypoint of the SBProxy. Run boots the application.
+func (s SBProxy) Run() {
+	var err error
+	ctx, cancel := context.WithCancel(context.Background())
+	defer waitWithTimeout(&group, time.Duration(s.AppConfig.TimeoutSec)*time.Second)
+	defer cancel()
+
+	handleInterrupts(ctx, cancel)
+
+	s.CronScheduler.Start()
+	defer s.CronScheduler.Stop()
+
+	addr := ":" + strconv.Itoa(s.AppConfig.Port)
+
+	logrus.Info("Running SBProxy server...")
+	if s.AppConfig.TLSKey != "" && s.AppConfig.TLSCert != "" {
```

Comment:
do we need tls? the sbproxy's architecture suggests that it will be called only by the platform it's running in

Reply:
if the proxy is registered as broker in k8s and the broker is accessed with tls, we may need it